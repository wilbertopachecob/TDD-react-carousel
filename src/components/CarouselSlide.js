import React from 'react';
import PropTypes from 'prop-types';

const CarouselSlide = ({ imgUrl, attribution, description, ...rest }) => (
  <figure {...rest}>
    <img src={imgUrl} />
    <figcaption>
      <strong>{description}</strong>
      {attribution}
    </figcaption>
  </figure>
);

CarouselSlide.propTypes = {
  imgUrl: PropTypes.string.isRequired,
  attribution: PropTypes.string,
  description: PropTypes.string,
};

export default CarouselSlide;
