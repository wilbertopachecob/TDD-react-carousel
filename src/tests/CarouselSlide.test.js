import { mount, shallow } from 'enzyme';
import CarouselSlide from '../components/CarouselSlide';
import React from 'react';

describe('CarouselSlide', () => {
  let wrapper;
  const imgUrl = 'https://example.com/image.png';

  beforeEach(() => {
    wrapper = shallow(<CarouselSlide imgUrl={imgUrl} />);
  });

  it('should render proper element', () => {
    expect(wrapper.type()).toBe('figure');
  });

  it('should render an <img> and <figcaption> element', () => {
    expect(wrapper.childAt(0).type()).toBe(CarouselSlide.defaultProps.Img);
    expect(wrapper.childAt(1).type()).toBe('figcaption');
  });

  it('passes `imgUrl` through to the <img>', () => {
    const Img = wrapper.find(CarouselSlide.defaultProps.Img);

    expect(Img.prop('src')).toBe(imgUrl);
  });

  it('uses `description` and `attribution` as the <figcaption>', () => {
    const description = 'A jaw-droppingly spectacular image';
    const attribution = 'Trevor Burnham';

    wrapper.setProps({ description, attribution });

    const figcaption = wrapper.find('figcaption');
    expect(figcaption.text()).toBe(`${description} ${attribution}`);
  });

  it('passes other props through to the <figure>', () => {
    const style = {};
    const onClick = () => {};
    const className = 'my-carousel-slide';

    wrapper.setProps({ style, onClick, className });

    expect(wrapper.prop('style')).toBe(style);
    expect(wrapper.prop('onClick')).toBe(onClick);
    expect(wrapper.prop('className')).toBe(className);
  });

  describe('Img', () => {
    let mounted;
    const Img = CarouselSlide.defaultProps.Img;
    const imgUrl = 'https://example.com/slide1.png';
    
    beforeEach(() => {
      mounted = mount(<Img src={imgUrl} imgHeight={500} />);
    });

    it('should display the img with the right src attr', () => {
      expect(mounted.containsMatchingElement(<img src={imgUrl} />)).toBe(true)
    });

    it('should display the img with the right properties', () => {
      expect(mounted).toHaveStyleRule('width', '100%');
    });

    it('uses imgHeight as the height style property', () => { 
      expect(mounted).toHaveStyleRule('height', '500px');
      mounted.setProps({imgHeight: 'calc(100vh - 100px)'});
      expect(mounted).toHaveStyleRule('height', 'calc(100vh - 100px)');
    });
  })
});
