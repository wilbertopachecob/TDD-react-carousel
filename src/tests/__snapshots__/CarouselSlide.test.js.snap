// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselSlide Img renders correctly 1`] = `
.c0 {
  object-fit: cover;
  width: 100%;
  height: 500px;
}

<img
  className="c0"
  src="https://example.com/slide1.png"
/>
`;

exports[`CarouselSlide renders correctly 1`] = `
<figure>
  <CarouselSlide__defaultImg
    imgHeight={500}
    src="https://example.com/image.png"
  />
  <figcaption>
    <strong>
      Description
       
    </strong>
    Attribution
  </figcaption>
</figure>
`;
